### DISCUSSION TITLE
How to include braces only when "more than one"

### DISCUSSION THREAD

**Discussion by asmaloney** [alecthomas/participle#199] [Q&A] [answered]
I have something like the following:

```
Foo  *foo   `parser:"( @@ "`
FooList []*foo `parser:"| '{' @@+ '}')"`
```

This works, but it's a bit messy to handle the `Foo` and `FooList` cases later on - I would prefer to just work with a `FooList`.

Is there a way to allow "one foo without braces BUT require braces if there is more than one"?

I tried what seems to be the obvious:
```
FooList []*foo `parser:"@@ | '{' @@+ '}'"`
```
but that doesn't work. I assume because `@@` won't map to a slice.

**Comment by alecthomas** [comment]
That construct definitely does work, so there must be some other issue. It could be ambiguity in your grammar. Try putting the `'{'` case first so there's no lookahead required.

**Comment by alecthomas** [comment]
But without an error or way to reproduce I can't be sure.

**Comment by asmaloney** [comment]
OK, thanks. If that's supposed to work I'll have to work on it some more to see what's going on.

I had tried switching the order as well, but I'm just getting a non-obvious parse error. I'm not sure how collapsing my two (working) patterns into one could result in ambiguity.

**Comment by asmaloney** [comment, marked as answer]
Ah, well. I was missing parens in the pattern. So it needs to look like this:

```
FooList []*foo `parser:"( '{' @@+ '} | @@ )'"`
```

That took far longer than it should have 😄
